File: dht11_thing.py
import Adafruit_DHT
from webthing import (Property, Thing, Value)

def update_values(sensor, pin, humidity_value, temperature_value):
    humidity, temperature = Adafruit_DHT.read_retry(sensor, pin)
    if humidity is not None and temperature is not None:
        humidity_value.notify_of_external_update(humidity)
        temperature_value.notify_of_external_update(temperature)

def make_dht11_thing(pin):
        humidity_thing = Thing(
            'urn:dev:ops:my-humidity-sensor-1234',
            'My Humidity Sensor',
            ['MultiLevelSensor'],
            'A web connected humidity sensor'
        )
        temperature_thing = Thing(
            'urn:dev:ops:my-temperature-sensor-1234',
            'My Temperature Sensor',
            ['MultiLevelSensor'],
            'A web connected temperature sensor'
        )
        humidity_value = Value(0.0)
        temperature_value = Value(0.0)
        humidity_thing.add_property(
            Property(humidity_thing,
                     'level',
                     humidity_value,
                     metadata={
                         '@type': 'LevelProperty',
                         'title': 'Humidity',
                         'type': 'number',
                         'description': 'The current humidity in %',
                         'minimum': 0,
                         'maximum': 100,
                         'unit': 'percent',
                         'readOnly': True,
                     }))
        temperature_thing.add_property(
            Property(temperature_thing,
                     'level',
                     temperature_value,
                     metadata={
                         '@type': 'LevelProperty',
                         'title': 'Temperature',
                         'type': 'number',
                         'description': 'The current temperature in °C',
                         'minimum': -5,
                         'maximum': 50,
                         'unit': 'celsius',
                         'readOnly': True,
                     }))
        sensor = Adafruit_DHT.DHT11
        return (humidity_thing, temperature_thing, lambda: update_values(sensor, pin, humidity_value, temperature_value))

File: dht11_thing.js
'use strict'

const {
  Property,
  Thing,
  Value
} = require('webthing')
const dht = require('node-dht-sensor')

function read (pin, humidityValue, temperatureValue) {
  let shouldStop = false
	const stop = () => shouldStop = true
  return function r() {
	if (shouldStop) {
	        return
	}
    dht.read(11, pin, function (error, temperature, humidity) {
      if (error) {
        console.log('error', error)
        setTimeout(r, 2000)
        return
      }
      console.log(`temp: ${temperature}°C, humidity: ${humidity}%`)
      humidityValue.notifyOfExternalUpdate(humidity)
      temperatureValue.notifyOfExternalUpdate(temperature)
      setTimeout(r, 2000)
    })
	  return stop
  }
}

// pin: physical pin
function makeDHTThing (pin) {
  const humidityThing = new Thing(
    'urn:dev:ops:kitchen-humidity-sensor-1234',
    'kitchen humidity',
    ['MultiLevelSensor'],
    'A web connected humidity sensor'
  )
  const temperatureThing = new Thing(
    'urn:dev:ops:kitchen-temperature-sensor-1234',
    'kitchen temperature',
    ['MultiLevelSensor'],
    'A web connected temperature sensor'
  )
  const humidityValue = new Value(0.0)
  const temperatureValue = new Value(0.0)
  humidityThing.addProperty(
    new Property(humidityThing,
      'level',
      humidityValue,
      {
        '@type': 'LevelProperty',
        title: 'Humidity',
        type: 'number',
        description: 'The current humidity in %',
        minimum: 0,
        maximum: 100,
        unit: 'percent',
        readOnly: true
      }))
  temperatureThing.addProperty(
    new Property(temperatureThing,
      'level',
      temperatureValue,
      {
        '@type': 'LevelProperty',
        title: 'Temperature',
        type: 'number',
        description: 'The current temperature in °C',
        minimum: -5,
        maximum: 40,
        unit: 'celsius',
        readOnly: true
      }))
  const stop = read(pin, humidityValue, temperatureValue)()
  return { humidityThing, temperatureThing, stop }
}

module.exports = { makeDHTThing }
